SetupDataPkg/ConfProfileMgrDxe/ConfProfileMgrDxe.c

```diff
@@ -0,0 +1,314 @@
+
+  @retval  EFI_SUCCESS   The driver validated the profile against flash contents successfully.
+  @retval  !EFI_SUCCESS  The driver failed to validate the profile against flash contents successfully.
+**/
+EFI_STATUS
+EFIAPI
+ConfProfileMgrDxeEntry (
+  IN  EFI_HANDLE        ImageHandle,
+  IN  EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_GUID    ActiveProfileGuid;
+  EFI_STATUS  Status;
+  EFI_GUID    CachedProfile;
+  UINT32      i;
+  UINT32      NumProfiles        = 0;
+  EFI_GUID    *ValidGuids        = NULL;
+  UINTN       Size               = sizeof (EFI_GUID);
+  UINT32      Attributes         = EFI_VARIABLE_NON_VOLATILE | EFI_VARIABLE_BOOTSERVICE_ACCESS;
+  BOOLEAN     FoundProfileInList = FALSE;
+  BOOLEAN     FoundCachedProfile = FALSE;
+
+  // read the cached profile value. This will be used in case RetrieveActiveProfileGuid fails
+  // and to check if we need to write the variable if the cached value is different than the new value
+  Status = gRT->GetVariable (
+                  CACHED_CONF_PROFILE_VARIABLE_NAME,
+                  &gConfProfileMgrVariableGuid,
+                  &Attributes,
+                  &Size,
+                  &CachedProfile
+                  );
+
+  if (EFI_ERROR (Status) || (Size != sizeof (EFI_GUID))) {
+    // failed to read the cached profile, fail back to generic profile if required
+    DEBUG ((DEBUG_WARN, "%a failed to read cached profile, expected on first boot (%r)!\n", __FUNCTION__, Status));
+    Size = sizeof (EFI_GUID);
+    CopyMem (&CachedProfile, (EFI_GUID *)&gSetupDataPkgGenericProfileGuid, Size);
+  } else {
+    // used to tell if we need to write the cached profile later or not
+    FoundCachedProfile = TRUE;
+  }
+
+  // RetrieveActiveProfileGuid will validate the profile guid is valid before returning
+  Status = RetrieveActiveProfileGuid (&ActiveProfileGuid);
+
+  // Check if the status failed, but don't error out, try to use cached profile
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a RetrieveActiveProfileGuid failed (%r)! Attempting to use cached profile\n", __FUNCTION__, Status));
+
+    if (!FoundCachedProfile) {
+      DEBUG ((DEBUG_ERROR, "%a Failed to retrieve cached profile, using generic profile\n", __FUNCTION__, Status));
+    }
+
+    CopyMem (&ActiveProfileGuid, &CachedProfile, Size);
+  }
+
+  // Validate this profile is in the list of profiles for this platform
+  NumProfiles = PcdGetSize (PcdConfigurationProfileList);
+
+  // if we get a bad size for the profile list, we need to fail to the default profile
+  // as we cannot validate the received profile
+  if ((NumProfiles == 0) || (NumProfiles % Size != 0)) {
+    DEBUG ((DEBUG_ERROR, "%a Invalid number of bytes in PcdConfigurationProfileList: %u, using generic profile\n", __FUNCTION__, NumProfiles));
+    ASSERT (FALSE);
+    CopyMem (&ActiveProfileGuid, (EFI_GUID *)&gSetupDataPkgGenericProfileGuid, Size);
+  } else {
+    NumProfiles /= (UINT32)Size;

```

Comment:
This is fine as is, but it is necessary to use Size instead of sizeof (EFI_GUID)?

Reply:
No reason to, I can switch it to sizeof.